sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects/objects.scala

```diff
@@ -885,13 +887,15 @@ case class MapObjects private(
         )
     }
 
-    // Make a copy of the data if it's unsafe-backed
-    def makeCopyIfInstanceOf(clazz: Class[_ <: Any], value: String) =
-      s"$value instanceof ${clazz.getSimpleName}? ${value}.copy() : $value"
+    // Make a copy of the unsafe data if the result contains any
+    def makeCopyUnsafeData(dataType: DataType, value: String) = {
+      s"""${value}.copyUnsafeData("${dataType.catalogString}")"""
```

Comment:
Is this safe for codgen? For complex type, will we bump into a very long string here?

Reply:
I guess it depends on how long would we classify as very long string. JSON is much easier to possibly create very long string, catalog string would create less. I agree the approach is questionable, but even getting a value from field requires schema for InternalRow so I'm not sure there's alternative approach for this.

Reply:
We may be able to let CodeGenerator go through the schema and generate code based on the schema, but I'd suspect the overall amount of generated code could be larger as generated code cannot leverage schema information and we have to generate all the codes accessing and replacing fields recursively.

And looks like string literal has 64k limit, which doesn't seem to be short - it's also the limitation of method length as well. We may be even be able to try to apply trick here, define this in static field or field of class, and assign to one String instance via splitting by chunk (less than 64k) and concatenating to get over the limitation.

Reply:
So the approach depends on how long of catalog string we are expecting for very complex type. 

1) If we expect this as not considerable size, it would work as it is. We might be able to refine a bit as defining local variable and assign from there, but it may require additional conditional logic.
2) If we expect less than 64k but considerable size and we don't want to let it consume the length of method limitation, we may need to define it to be one of static fields.
3) If we expect more than 64k, we should have it to be one of static fields, and also break catalog string down to chunk, and concat to one of String instance.

Do we have some huge and complicated but realistic (in production) schema to test this?